src/components/RichContenteditable/RichContenteditable.vue

```diff
@@ -0,0 +1,328 @@
+		* @param {string} search the query
+		* @param {Function} callback the callback to process the results with
+		*/
+		autoComplete(search, callback) {
+			return callback(Object.values(this.userData))
+		}
+	}
+}
+</script>
+```
+
+</docs>
+
+<template>
+	<div ref="contenteditable"
+		:class="{'rich-contenteditable__input--empty': isEmptyValue}"
+		:contenteditable="contenteditable"
+		:placeholder="placeholder"
+		class="rich-contenteditable__input"
+		@input="onInput"
+		@keydown.delete="onDelete"
+		@paste.stop.prevent="onPaste" />
+</template>
+
+<script>
+import Tribute from 'tributejs/dist/tribute.esm'
+import debounce from 'debounce'
+
+import { t } from '../../l10n.js'
+import AutoCompleteResult from './AutoCompleteResult'
+import richEditor from '../../mixins/richEditor/index'
+
+export default {
+	name: 'RichContenteditable',
+
+	mixins: [richEditor],
+
+	props: {
+		value: {
+			type: String,
+			required: true,
+		},
+		placeholder: {
+			type: String,
+			default: t('Write message, @ to mention someone …'),
+		},
+		autoComplete: {
+			type: Function,
+			required: true,
+		},
+		menuContainer: {
+			type: Element,
+			default: () => document.body,
+		},
+		/**
+		 * Is the content editable
+		 */
+		contenteditable: {
+			type: Boolean,
+			default: true,
+		},
+	},
+
+	data() {
+		return {
+			tribute: null,
+			options: {
+				fillAttr: 'id',
+				// Search against id and label (display name)
+				lookup: result => `${result.id} ${result.label}`,
+				// Where to inject the menu popup
+				menuContainer: this.menuContainer,
+				// Popup mention autocompletion templates
+				menuItemTemplate: item => this.renderComponentHtml(item.original, AutoCompleteResult),
+				// Hide if no results
+				noMatchTemplate: () => '<span class="hidden"></span>',
+				// Inner display of mentions
+				selectTemplate: item => this.genSelectTemplate(item.original?.id),
+				// Autocompletion results
+				values: this.debouncedAutoComplete,
+			},
+
+			// Represent the raw untrimmed text of the contenteditable
+			// serves no other purpose than to check whether the
+			// content is empty or not
+			localValue: '',
+		}
+	},
+
+	computed: {
+		isEmptyValue() {
+			return !this.localValue || this.localValue.trim() === ''
+		},
+
+		/**
+		 * Is this Firefox? 🙄
+		 * @returns {boolean}
+		 */
+		isFF() {
+			return !!navigator.userAgent.match(/firefox/i)
+		},
+
+	},
+
+	watch: {
+		/**
+		 * If the parent value change, we compare the plain text rendering
+		 * If it's different, we render everything and update the main content
+		 */
+		value() {
+			const html = this.$refs.contenteditable.innerHTML
+			// Compare trimmed versions to be safe
+			if (this.value.trim() !== this.parseContent(html).trim()) {
+				const renderedContent = this.renderContent(this.value)
+				this.$refs.contenteditable.innerHTML = renderedContent
+				this.localValue = this.value
+			}
+		},
+	},
+
+	mounted() {
+		this.tribute = new Tribute(this.options)
```

Comment:
The styling is handled there, not in apps. Thus needs to be global.
I don't think that is necessary then :)

Reply:
Of course, the main Tribute style should be the same across apps, so it is fine that it is global. But apps can specify the icons of the autocompletion candidates, so in some cases it may be necessary that the Tribute container element has class `tribute-container talk` rather than just `tribute-container` (that is what I meant by customizing the class; not to change the CSS rules of the global tribute container, but to apply CSS rules to the tribute container results only when they belong to a specific app).

Let me try to provide a (contrived :-P ) example:
- Talk shows "all" mentions with a Talk icon (not true, it uses a group icon, but it is an example :-P )
- Comments shows "all" mentions with a Comments icon (same :-P )
- The Tribute container is a direct child of the body, so CSS rules for `.tribute-container .icon-all` need to be applied globally (I mean, they can not, as far as I know, be included in the style of the Vue component using RichContenteditable). Therefore, Comments and Talk both apply their own rule to show the desired background-image in `.tribute-container .icon-all`
- When Talk is loaded in the Files app its global CSS rule for `.tribute-container .icon-all` overrides the one previously defined by the Comments app
- When the user types `@all` in the Comments app the icon in the autocompletion popup shows Talk rather than Comments

If apps can specify an additional CSS class to be added to the Tribute container element then there will be no clash, as they could specify the icons for `.tribute-container.talk .icon-all` and `.tribute-container.comments .icon-all`.

When using atwho it was really needed, because the autocompletion popup for Talk and the Comments app had different styles that clashed with each other. With Tribute the problem should be limited to icons, and it will be much less likely that the same icon name in two different apps has two different images. But it is something to keep in mind :-)

Reply:
It doesn't matter, the @all icon is provided by the AutoComplete dataset.
Don't pass an icon-all, pass the appropriate icon string directly: `icon-public` `icon-groups`... etc etc

The icons are global,  you don't need to style the tribute container. Just don't pass the same data to the RichDocumenteditor depending on the context.